### PR TITLE
feat: Handle sum types in backend encoder

### LINKED ISSUE #1387: Handle sum types in backend encoder
Use the [type registry](#1386)

### PR DESCRIPTION
**Pull request by worstell**
fixes #1387

### CODE REVIEW

File: go-runtime/encoding/encoding.go [hunk 1]
```diff
@@ -93,18 +95,29 @@ func encodeValue(v reflect.Value, w *bytes.Buffer) error {
...
+				stv := struct {
+					Name  string
+					Value any
+				}{Name: vName, Value: v.Elem().Interface()}
+				return encodeValue(ctx, reflect.ValueOf(stv), w)
+			}
 		}
-		return encodeValue(v.Elem(), w)
+
+		return fmt.Errorf("the only interface types are enums or any, not %s", t)
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:113]:
```suggestion
		return fmt.Errorf("the only supported interface types are enums or any, not %s", t)
```

File: go-runtime/encoding/encoding.go [hunk 2]
```diff
@@ -356,6 +375,71 @@ func decodeMap(d *json.Decoder, v reflect.Value) error {
 	return err
 }
 
+func decodeSumType(ctx context.Context, d *json.Decoder, v reflect.Value) error {
+	tr, ok := typeregistry.FromContext(ctx).Get()
+	if !ok {
+		return fmt.Errorf("no type registry found in context")
+	}
+
+	if err := expectDelim(d, '{'); err != nil {
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:384]: It should be possible to use `json.Decoder.Decode` directly here no, and avoid all this manual decoding? eg. something like [this](https://go.dev/play/p/EgA1MXlH35C)

File: go-runtime/ftl/typeregistry/type_registry.go [hunk 3]
```diff
@@ -13,46 +14,88 @@ func ContextWithTypeRegistry(ctx context.Context, r *TypeRegistry) context.Conte
...
+	}
+	return optional.None[*TypeRegistry]()
+}
+
+// TypeRegistry is used for dynamic type resolution at runtime. It stores associations between sum type discriminators
+// and their variants, for use in encoding and decoding.
 //
-// FTL manages the type registry for you, so you don't need to create one yourself
+// FTL manages the type registry for you, so you don't need to create one yourself.
 type TypeRegistry struct {
```
Comment (alecthomas) [inline go-runtime/ftl/typeregistry/type_registry.go:31]: Nice

File: go-runtime/ftl/typeregistry/type_registry.go [hunk 4]
```diff
@@ -13,46 +14,88 @@ func ContextWithTypeRegistry(ctx context.Context, r *TypeRegistry) context.Conte
...
-	if !ok {
-		return nil, fmt.Errorf("type %q not registered", name)
+		sumTypes: make(map[string][]sumTypeVariant),
 	}
-	return reflect.New(typ).Interface(), nil
 }
 
 // RegisterSumType registers a Go sum type with the type registry. Sum types are represented as enums in the
 // FTL schema.
 func (t *TypeRegistry) RegisterSumType(discriminator reflect.Type, variants map[string]reflect.Type) {
```
Comment (alecthomas) [inline go-runtime/ftl/typeregistry/type_registry.go:50]: What are they keys to `variants`? The unqualified FTL type names, Go type names?

**Review by alecthomas — APPROVED** [review]
Nice!

File: go-runtime/encoding/encoding.go (hunk 2 above)
Comment (worstell) [reply go-runtime/encoding/encoding.go:384]: totally 🤦‍♀️

File: go-runtime/ftl/typeregistry/type_registry.go (hunk 4 above)
Comment (worstell) [reply go-runtime/ftl/typeregistry/type_registry.go:50]: the keys are the variants names, e.g.:
```
typeRegistry.RegisterSumType(reflect.TypeFor[foo.TypeEnum](), map[string]reflect.Type{
		"A": reflect.TypeFor[foo.A](),
		"B": reflect.TypeFor[foo.B](),
})
```

fwiw this part is only used by generated code!

File: go-runtime/encoding/encoding_test.go [hunk 5]
```diff
@@ -85,13 +103,21 @@ func TestUnmarshal(t *testing.T) {
 			Bool   bool
 		}{ftl.None[int](), true}},
 		{name: "Pointer", input: `{"string":"foo"}`, expected: &struct{ String string }{}, err: `pointer types are not supported: *struct { String string }`},
+		{name: "SumType", input: `{"d":{"name":"Variant","value":{"message":"hello"}}}`, expected: struct{ D discriminator }{variant{"hello"}}},
+		{name: "MalformedSumType", input: `{"d":{"message":"hello"}}`, expected: struct{ D discriminator }{}, err: `unexpected key "message" in sum type variant, expected "name" or "value"`},
```
[the remark this answers is not in the document]
Comment (worstell) [reply go-runtime/encoding/encoding_test.go:107]: will do!

**Pull request merged**